File: 739. Daily Temperatures/main.py
from typing import List


class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        tempSize = len(temperatures)

        result = [0] * tempSize
        stack = []

        for i in range(tempSize):
            while stack and temperatures[i] > temperatures[stack[-1]]:
                index = stack.pop()
                result[index] = i - index

            stack.append(i)

        return result

File: 739. Daily Temperatures/main.rs
struct Solution;

impl Solution {
    pub fn daily_temperatures(temperatures: Vec<i32>) -> Vec<i32> {
        let temp_size = temperatures.len();
        let mut result = vec![0; temp_size];
        let mut stack = Vec::new();

        for i in 0..temp_size {
            while !stack.is_empty() && temperatures[i] > temperatures[*stack.last().unwrap()] {
                let index = stack.pop().unwrap();
                result[index] = (i - index) as i32;
            }

            stack.push(i);
        }

        result
    }
}


impl Solution {
    pub fn daily_temperatures(temperatures: Vec<i32>) -> Vec<i32> {
        let temp_size = temperatures.len();
        let mut result = vec![0; temp_size];
        let mut stack = Vec::new();

        for i in 0..temp_size {
            while let Some(&index) = stack.last() {
                if temperatures[i] <= temperatures[index] {
                    break;
                }
                stack.pop();
                result[index] = (i - index) as i32;
            }

            stack.push(i);
        }

        result
    }
}
